beacon-chain/blockchain/forkchoice/tree_test.go

```diff
@@ -0,0 +1,144 @@
+package forkchoice
+
+import (
+	"context"
+
+	"github.com/prysmaticlabs/go-ssz"
+	"github.com/prysmaticlabs/prysm/beacon-chain/db"
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	ethpb "github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+)
+
+// blockTree1 constructs the following tree:
+//    /- B1
+// B0           /- B5 - B7
+//    \- B3 - B4 - B6 - B8
+// (B1, and B3 are all from the same slots)
+func blockTree1(db db.Database) ([][]byte, error) {
+	b0 := &ethpb.BeaconBlock{Slot: 0, ParentRoot: []byte{'g'}}
+	r0, _ := ssz.SigningRoot(b0)
+	b1 := &ethpb.BeaconBlock{Slot: 1, ParentRoot: r0[:]}
+	r1, _ := ssz.SigningRoot(b1)
+	b3 := &ethpb.BeaconBlock{Slot: 3, ParentRoot: r0[:]}
+	r3, _ := ssz.SigningRoot(b3)
+	b4 := &ethpb.BeaconBlock{Slot: 4, ParentRoot: r3[:]}
+	r4, _ := ssz.SigningRoot(b4)
+	b5 := &ethpb.BeaconBlock{Slot: 5, ParentRoot: r4[:]}
+	r5, _ := ssz.SigningRoot(b5)
+	b6 := &ethpb.BeaconBlock{Slot: 6, ParentRoot: r4[:]}
+	r6, _ := ssz.SigningRoot(b6)
+	b7 := &ethpb.BeaconBlock{Slot: 7, ParentRoot: r5[:]}
+	r7, _ := ssz.SigningRoot(b7)
+	b8 := &ethpb.BeaconBlock{Slot: 8, ParentRoot: r6[:]}
+	r8, _ := ssz.SigningRoot(b8)
+	for _, b := range []*ethpb.BeaconBlock{b0, b1, b3, b4, b5, b6, b7, b8} {
+		if err := db.SaveBlock(context.Background(), b); err != nil {
+			return nil, err
+		}
+		if err := db.SaveState(context.Background(), &pb.BeaconState{}, bytesutil.ToBytes32(b.ParentRoot)); err != nil {
+			return nil, err
+		}
+	}
+	return [][]byte{r0[:], r1[:], nil, r3[:], r4[:], r5[:], r6[:], r7[:], r8[:]}, nil
+}
+
+// blockTree2 constructs the following tree:
+// Scenario graph: shorturl.at/loyP6
+//
+//digraph G {
+//    rankdir=LR;
+//    node [shape="none"];
+//
+//    subgraph blocks {
+//        rankdir=LR;
+//        node [shape="box"];
+//        a->b;
+//        a->c;
+//        b->d;
+//        b->e;
+//        c->f;
+//        c->g;
+//        d->h
+//        d->i
+//        d->j
+//        d->k
+//        h->l
+//        h->m
+//        g->n
+//        g->o
+//        e->p
+//    }
+//}
+func blockTree2(db db.Database) ([][]byte, error) {
+	b0 := &ethpb.BeaconBlock{Slot: 0, ParentRoot: []byte{'g'}}
```

Comment:
Can this be constructed in a better way? Seems like a nightmare to maintain or edit down the line

Reply:
it'll be hard to construct such tree w/o manually editing it, the good thing is we have a clear Scenario graph using digraph.  I also have a yaml driven test frame coming as part 3: https://github.com/prysmaticlabs/prysm/pull/3213

it'll be easier to define a test tree there